Libplanet.Extensions.Cocona/Commands/StoreQueryCommand.cs

```diff
@@ -0,0 +1,142 @@
+using System;
+using System.Collections.Immutable;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Bencodex.Types;
+using Cocona;
+using Libplanet.Action;
+using Libplanet.Blocks;
+using Libplanet.Tx;
+
+namespace Libplanet.Extensions.Cocona.Commands
+{
+    public class StoreQueryCommand
+    {
+        [Command(Description = "Query block by the index")]
+        public void BlockByIndex(
+            [Argument("HOME", Description = "root home path")]
+            string home,
+            [Argument("BLOCK-INDEX", Description = "block index")]
+            int blockIndex
+        )
+        {
+            var store = new RocksDBStore.RocksDBStore(home);
+            var chainId = store.GetCanonicalChainId();
+            if (chainId == null)
+            {
+                Console.Error.WriteLine("cannot find the main branch of the blockchain");
+                return;
+            }
+
+            BlockHash? blockHash = store.IndexBlockHash(chainId.Value, blockIndex);
+            if (blockHash == null)
+            {
+                Console.Error.WriteLine($"cannot find the block with the height[{blockIndex}]" +
+                                        $" within the blockchain[{chainId}]");
+                return;
+            }
+
+            Block<DummyAction> block = store.GetBlock<DummyAction>(blockHash.Value);
+            if (block == null)
+            {
+                Console.Error.WriteLine($"cannot find the block with the hash[{blockHash}]");
+            }
```

Comment:
When you need get `T` from `T?`, you'd better do pattern matching:

```csharp
T? nullableValue = GetNullableValue();
if (nullableValue is { } v)  // Or it can be like: (nullableValue is T v)
{
    // Inside this, the of `v` is not `T?` but `T`.
}
```

If you need to return early:

```csharp
T? nullableValue = GetNullableValue();
if (!(nullableValue is { } v))  // Or: (!(nullableValue is T v))
{
    return;
}
// From this, the of `v` is not `T?` but `T`.
```

---

If the command is not successful, the exit code should be non-zero (according to the Unix tradition).  Cocona, the CLI framework that we use, provides `CommandExitedException` class for non-zero exit code:

```csharp
if (unsuccessful)
{
    // Note that the below function does not throw an exception, but just returns it.
    // You should throw it by yourself:
    throw Utils.Error("Something went wrong.");
}
```

However, we already have the [`Utils.Error(string, int)` utility function](https://github.com/planetarium/libplanet/blob/0a5702ca57612ef5275a7e8ad88bf21ad38474b9/Libplanet.Extensions.Cocona/Utils.cs#L11-L16) for this:

```csharp
if (unsuccessful)
{
    Utils.Error("Something went wrong.", 1);
    throw new CommandExitedException(1);
}
```

---

To put them into this code:

```suggestion
            if (!(store.GetCanonicalChainId() is Guid chainId))
            {
                throw Utils.Error("Cannot find the main branch of the blockchain.");
            }

            if (!(store.IndexBlockHash(chainId, blockIndex) is BlockHash blockHash))
            {
                throw Utils.Error(
                    $"Cannot find the block with the height {blockIndex}" +
                    $" within the blockchain {chainId}."
                );
            }

            if (!(store.GetBlock<DummyAction>(blockHash) is Block<DummyAction> block))
            {
                throw Utils.Error($"Cannot find the block with such hash: {blockHash}.");
            }
```

Reply:
```csharp
T? nullableValue = GetNullableValue();
if (nullableValue is { } v)  // Or it can be like: (nullableValue is T v)
{
    // Inside this, the of `v` is not `T?` but `T`.
}

```
- what `{ }` mean in this context?
- `v` can be accessible outside of the if clause. but [ms the document](https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching) says that it can be accessed only inside the if clause. The doc is wrong. right?

Reply:
- `{ }` is an empty [property pattern][1], which matches to any non-null objects/values.
- If you negate (`!(...)`) the pattern, it cannot be used *inside* the block, but only can be *outside* the block.  (FYI C# 9 will support `is not` operator for this use case, but `!(...)` already has the same functionality even though it looks quite ugly.)

[1]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#property-pattern

Reply:
Understood. Thanks!